apps/astarte_trigger_engine/lib/astarte_trigger_engine/amqp_consumer/amqp_message_consumer.ex

```diff
@@ -0,0 +1,236 @@
+
+  @impl true
+  def handle_info(:connect, state), do: do_connect(state)
+
+  @impl true
+  def handle_info(
+        {:DOWN, monitor, :process, chan_pid, _reason},
+        %{monitor: monitor, channel: %{pid: chan_pid}} = state
+      ) do
+    do_connect(state)
+  end
+
+  # Confirmation sent by the broker after registering this process as a consumer
+  @impl true
+  def handle_info(
+        {:basic_consume_ok, _consumer_tag},
+        %{realm_name: realm_name, policy: policy} = state
+      ) do
+    _ =
+      Logger.debug(
+        "consumer for #{realm_name}, #{policy.name} successfully registered",
+        tag: "basic_consume_ok"
+      )
+
+    {:noreply, state}
+  end
+
+  # This is sent for each message consumed, where `payload` contains the message
+  # content and `meta` contains all the metadata set when sending with
+  # Basic.publish or additional info set by the broker;
+  @impl true
+  def handle_info(
+        {:basic_deliver, payload, meta},
+        %{realm_name: realm_name, policy: policy, channel: chan} = state
+      ) do
+    _ =
+      Logger.debug(
+        "consumer for #{realm_name}, #{policy.name} received message, payload: #{inspect(payload)}, meta: #{
+          inspect(meta)
+        }",
+        tag: "message_received"
+      )
+
+    get_policy_process(realm_name, policy)
+    |> Policy.handle_event(chan, payload, meta)
+
+    {:noreply, state}
+  end
+
+  # Sent by the broker when the consumer is unexpectedly cancelled (such as after a queue deletion)
+  @impl true
+  def handle_info({:basic_cancel, _}, %{realm_name: realm_name, policy: policy} = state) do
+    _ =
+      Logger.error("consumer for #{realm_name}, #{policy.name} was cancelled by the broker",
+        tag: "basic_cancel"
+      )
+
+    {:stop, :shutdown, state}
+  end
+
+  # Confirmation sent by the broker to the consumer process after a Basic.cancel
+  @impl true
+  def handle_info({:basic_cancel_ok, _}, %{realm_name: realm_name, policy: policy} = state) do
+    _ =
+      Logger.info("consumer for #{realm_name}, #{policy.name} was cancelled by the broker",
+        tag: "basic_cancel_ok"
+      )
+
+    {:stop, :normal, state}
+  end
+
+  defp schedule_connect() do
+    Process.send_after(self(), :connect, @reconnect_interval)
+  end
+
+  # Gets a connection worker out of the connection pool, if there is one available
+  # takes a channel out of it channel pool, if there is one available
+  # subscribe itself as a consumer process.
+  defp do_connect(state) do
+    ExRabbitPool.get_connection_worker(:events_consumer_pool)
+    |> ExRabbitPool.checkout_channel()
+    |> handle_channel_checkout(state)
+  end
+
+  # When successfully checks out a channel, sets up exchange and queue, subscribe itself as a consumer
+  # process and monitors it handle crashes and reconnections
+  defp handle_channel_checkout(
+         {:ok, %{pid: channel_pid} = channel},
+         %{policy: policy, realm_name: realm_name} = state
+       ) do
```

Comment:
Don't extract everything in the function head if it's not necessary for a guard (see [here](https://github.com/lucasvegi/Elixir-Code-Smells#complex-extractions-in-clauses))

Reply:
Also, see comment above, I'd rename this to the clearer `try_to_connect`